pkg/reconciler/volumeclaim/pvchandler.go

```diff
@@ -0,0 +1,74 @@
+package volumeclaim
+
+import (
+	"fmt"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	errorutils "k8s.io/apimachinery/pkg/util/errors"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+const (
+	// ReasonCouldntCreateWorkspacePVC indicates that a Pipeline expects a workspace from a
+	// volumeClaimTemplate but couldn't create a claim.
+	ReasonCouldntCreateWorkspacePVC = "CouldntCreateWorkspacePVC"
+)
+
+type PvcHandler interface {
+	CreatePersistentVolumeClaimsForWorkspaces(wb []v1alpha1.WorkspaceBinding, ownerReference metav1.OwnerReference, namespace string) error
+}
+
+type defaultPVCHandler struct {
+	clientset clientset.Interface
+	logger    *zap.SugaredLogger
+}
+
+func NewPVCHandler(clientset clientset.Interface, logger *zap.SugaredLogger) PvcHandler {
+	return &defaultPVCHandler{clientset, logger}
+}
+
+func (c *defaultPVCHandler) CreatePersistentVolumeClaimsForWorkspaces(wb []v1alpha1.WorkspaceBinding, ownerReference metav1.OwnerReference, namespace string) error {
+	var errs []error
+	for _, claim := range getPersistentVolumeClaims(wb, ownerReference, namespace) {
+		_, err := c.clientset.CoreV1().PersistentVolumeClaims(claim.Namespace).Get(claim.Name, metav1.GetOptions{})
+		switch {
+		case apierrors.IsNotFound(err):
+			_, err := c.clientset.CoreV1().PersistentVolumeClaims(claim.Namespace).Create(claim)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("failed to create PVC %s: %s", claim.Name, err))
+			}
+			if err == nil || !apierrors.IsAlreadyExists(err) {
+				c.logger.Infof("Created PersistentVolumeClaim %s in namespace %s", claim.Name, claim.Namespace)
+			}
+		case err != nil:
+			errs = append(errs, fmt.Errorf("failed to retrieve PVC %s: %s", claim.Name, err))
+		}
+	}
+	return errorutils.NewAggregate(errs)
+}
+
+func getPersistentVolumeClaims(workspaceBindings []v1alpha1.WorkspaceBinding, ownerReference metav1.OwnerReference, namespace string) map[string]*corev1.PersistentVolumeClaim {
+	claims := make(map[string]*corev1.PersistentVolumeClaim)
+	for _, workspaceBinding := range workspaceBindings {
+		if workspaceBinding.VolumeClaimTemplate == nil {
+			continue
+		}
+
+		claim := workspaceBinding.VolumeClaimTemplate.DeepCopy()
+		claim.Name = GetPersistentVolumeClaimName(workspaceBinding.VolumeClaimTemplate, workspaceBinding, ownerReference)
+		claim.Namespace = namespace
+		claim.OwnerReferences = []metav1.OwnerReference{ownerReference}
+		claims[workspaceBinding.Name] = claim
+	}
+	return claims
+}
+
+// getPersistentVolumeClaimName gets the name of PersistentVolumeClaim for a Workspace and PipelineRun or TaskRun. claim
+// must be a PersistentVolumeClaim from set's VolumeClaims template.
+func GetPersistentVolumeClaimName(claim *corev1.PersistentVolumeClaim, wb v1alpha1.WorkspaceBinding, owner metav1.OwnerReference) string {
```

Comment:
Is this guaranteed to be unique?  I'm trying to figure out the possible combinations and I have a hunch that there could be two Tasks in a Pipeline that share a PVC Claim Name, WorkspaceBinding Name, and PipelineRun (owner) Name.  In this case would there be potential for collision?

Reply:
Thanks for reviewing! There are two cases. I tried to explain them in my PR description under "How it works". When the volumeClaimTemplate is declared for a PipelineRun (owner) - it will be seen as a "PersistentVolumeClaim"-workspace for all TaskRuns in the Pipeline. The function `taskWorkspaceByWorkspaceVolumeSource()` in `pipelinerun.go` applies the template and pass it as PersistentVolumeClaim to the TaskRuns - this happens after the PVC is created from the template - after the `IsBeforeFirstTaskRun()`-check. I can explain this more in person if you want.

Reply:
PipelineRun has a unique name, and will be `owner.Name` here. There can also be multiple workspaces per PipelineRun but they must have unique names. Hence the PVC get an unique name here.